invenio_communities/communities/services/components.py

```diff
@@ -169,3 +172,78 @@ def create(self, identity, data=None, record=None, **kwargs):
 
         # Invalidate the membership cache
         on_membership_change(identity=identity)
+
+
+class OAISetComponent(ServiceComponent):
+    """Service component for OAI set integration."""
+
+    _oai_sets_prefix = "community"
+
+    def _retrieve_set(self, record, **kwargs):
+        return OAISet.query.filter(OAISet.spec == self._create_set_spec(record.get("id"))).first()
+
+    def _create_set_spec(self, community_id):
+        return "{prefix}-{id}".format(prefix=self._oai_sets_prefix, id=community_id)
+
+    def _create_set_description(self, community_title):
+        return "Records belonging to the community '{title}'".format(title=community_title)
+
+    def _create_set_from_community(self, record, **kwargs):
+        community_id = kwargs.get("id") or record.get("id")
+        community_title = record.metadata.get("title")
+
+        community_set = OAISet()
+        community_set.name = community_title
+        community_set.description = self._create_set_description(community_title)
+        community_set.spec = self._create_set_spec(community_id)
+        community_set.search_pattern = "parent.communities.ids:{id}".format(id=community_id)
+
+        return community_set
+
+    def create(self, identity, data=None, record=None, **kwargs):
+        """Create a set for the community."""
+        if record.get("access", {}).get("visibility") != "public":
+            return
+
+        community_set = self._create_set_from_community(record)
+        # NOTE: will be indexed via a listener in oaiserver module
+        db.session.add(community_set)
+
+
+    def delete(self, identity, data=None, record=None, **kwargs):
+        """Remove the set for the community."""
+        community_set = self._retrieve_set(record)
+        if not community_set:
+            return
+
+        # NOTE: will be removed from index via listener in oaiserver module
+        db.session.delete(community_set)
+
+
+    def update(self, identity, data=None, record=None, **kwargs):
+        """Update set accordingly."""
+        if record.get("access", {}).get("visibility") != "public":
+            # delete set if the community is not public anymore
+            self.delete(identity, data, record, **kwargs)
+            return
+
+        community_set = self._retrieve_set(record)
+        if not community_set:
+            # create new set because community is now public
+            self.create(identity, data, record, **kwargs)
+            return
+
+        # just update set with new name and description
+        community_set.name = record.metadata.get("title")
```

Comment:
Technically it should not be possible to have a community without a title so I think you can do ``record.metadata['title']``